File: src/BFSE.java
import java.util.*;
import java.io.*;

public class BFSE {
	
	private static final Integer INF = 0x0F_FF_FF_FF;
	private static final Integer NEG_INF = 0xF0_00_00_01;
	private ArrayList<ArrayList<Edge>> adjList;
	private ArrayList<Integer> distance;
	private ArrayList<Integer> parent;
	private int factor;
	private int absoluteMinimum;

	// O(1) initialization
	public BFSE() {
		super();
		
		this.adjList = new ArrayList<ArrayList<Edge>>();
		this.distance = new ArrayList<Integer>();
		this.parent = new ArrayList<Integer>();
		
		factor = 0;
		absoluteMinimum = 0;
	}
	
	// O((initial - V) + (terminal - V)) insertion
	public void addEdge(Edge edge){
		
		int initial = edge.getInitial();
		int terminal = edge.getTerminal();
		int weight = edge.getWeight();
		
		//factor set to minimum edge weight
		if(factor == 0)
			factor=weight;
		else if(factor>weight)
			factor = weight;
		
		if(weight<0)
			absoluteMinimum+=weight;
		
		//increase capacity to fit new initial vertex
		if(initial + 1>adjList.size()){
			int amountToAdd = initial + 1 - adjList.size();
			adjList.addAll(Collections.nCopies(amountToAdd, null));
			distance.addAll(Collections.nCopies(amountToAdd, INF));
			parent.addAll(Collections.nCopies(amountToAdd, -1));
		}
		
		//increase capacity to fit new terminal vertex
		if( terminal + 1 > adjList.size()){
			int amountToAdd = terminal + 1 - adjList.size();
			adjList.addAll(Collections.nCopies(amountToAdd, null));
			distance.addAll(Collections.nCopies(amountToAdd, INF));
			parent.addAll(Collections.nCopies(amountToAdd, -1));
		}		
		
		//initialize this vertex's edgelist if null
		if(adjList.get(initial)==null){
			adjList.set(initial, new ArrayList<Edge>());
		}
		
		//update the arrays
		adjList.get(initial).add(edge);
	}
	
	public void addEdge(int weight, int initial, int terminal){
		addEdge(new Edge( weight,  initial,  terminal));
	}

	// O(V+E*w(E))
	public ArrayList<Integer> sssp(int source){
		
		//copy over everything so we don't destroy the original graph
		ArrayList<Integer> ssspParent = new ArrayList<Integer>(parent);
		ArrayList<Integer> ssspDistance = new ArrayList<Integer>(distance);
		ArrayList<ArrayList<Edge>> ssspAdjList = new ArrayList<ArrayList<Edge>>(adjList);
		Queue<Object> queue = new LinkedList<Object>();
		
		//set source values
		ssspDistance.set(source, 0);
		queue.add(source);
		
		while(!queue.isEmpty()){
			
			Object item = queue.poll();
			int initial;
			Edge expanded;
			
			// if this is a vertex
			// explore the neighbors
			if(item instanceof Integer){
				initial = (Integer) item;
				Iterator<Edge> it;
				
				//if this vertex has outgoing edges
				if(ssspAdjList.get(initial)!=null){
					
					it = ssspAdjList.get(initial).iterator();
					
					//for each edge from this vertex
					while(it.hasNext()){
						
						//get the edge
						Edge edge = it.next();
						step(edge, queue, ssspParent, ssspDistance);
					}
				}
			
			//if this is an edge
			//increment the edge's step
			}else if(item instanceof Edge){
				
				expanded = (Edge) item;
				step(expanded, queue, ssspParent, ssspDistance);
			}
		}
		
		System.out.println("Distances: " + ssspDistance);
		System.out.println("Parents: " + ssspParent);
		System.out.println();
		
		return ssspDistance;
	}
	
	public void step(Edge edge, Queue<Object> queue, 
			ArrayList<Integer> ssspParent, 
			ArrayList<Integer> ssspDistance){
	
		int weight = edge.getWeight();
		int initial = edge.getInitial();
		int terminal = edge.getTerminal();
		int step = edge.getStep();
	
		//check to see if this 
		//edge is fully traversed
		if(step>factor){
			
			//if the edge is not 
			//traversed, then move
			//up edge one step
			//and add the edge 
			//to the queue
			edge.setStep(step-1);
			queue.add(edge);
		}
		//check to see if the
		//terminal vertex at this
		//edge has been visited
		else if(ssspDistance.get(terminal) > ssspDistance.get(initial) + weight){
		
			//check for negative cycle
			if(ssspDistance.get(terminal)<absoluteMinimum){
				ssspDistance.set(terminal, NEG_INF);
			}else{
				//relax this edge
				ssspDistance.set(terminal, ssspDistance.get(initial)+weight);
			}
			
			ssspParent.set(terminal,initial);
			
			//mark the terminal vertex
			//for exploration
			queue.add(terminal);
		}
	}

	@Override
	public String toString() {
		return "BFSE [adjList=" + adjList + "]";
	}
	
	
}

File: src/bfse.ts
interface Edge {
  weight: number;
  initial: number;
  terminal: number;
  step: number;
}

export class BFSE {
  private readonly INF = 0x0FFFFFFF;
  private readonly NEG_INF = -268435455;
  private adjList: Array<Array<Edge | null>>;
  private distance: number[];
  private parent: number[];
  private factor: number;
  private absoluteMinimum: number;

  constructor() {
    this.adjList = [];
    this.distance = [];
    this.parent = [];
    this.factor = 0;
    this.absoluteMinimum = 0;
  }

  addEdge(edge: Edge): void {
    const { initial, terminal, weight } = edge;

    if (this.factor === 0) {
      this.factor = weight;
    } else if (this.factor > weight) {
      this.factor = weight;
    }

    if (weight < 0) {
      this.absoluteMinimum += weight;
    }

    if (initial + 1 > this.adjList.length) {
      const amountToAdd = initial + 1 - this.adjList.length;
      this.adjList.push(...Array(amountToAdd).fill(null));
      this.distance.push(...Array(amountToAdd).fill(this.INF));
      this.parent.push(...Array(amountToAdd).fill(-1));
    }

    if (terminal + 1 > this.adjList.length) {
      const amountToAdd = terminal + 1 - this.adjList.length;
      this.adjList.push(...Array(amountToAdd).fill(null));
      this.distance.push(...Array(amountToAdd).fill(this.INF));
      this.parent.push(...Array(amountToAdd).fill(-1));
    }

    if (!this.adjList[initial]) {
      this.adjList[initial] = [];
    }

    this.adjList[initial].push(edge);
  }

  addEdgeWithParams(weight: number, initial: number, terminal: number): void {
    this.addEdge({ weight, initial, terminal, step: 0 });
  }

  sssp(source: number): number[] {
    const ssspParent = [...this.parent];
    const ssspDistance = [...this.distance];
    const ssspAdjList: any = this.adjList.map((edges) => (edges ? [...edges] : null));
    const queue: Array<number | Edge> = [];

    ssspDistance[source] = 0;
    queue.push(source);

    while (queue.length > 0) {
      const item = queue.shift()!;
      let initial: number;
      let expanded: Edge;

      if (typeof item === 'number') {
        initial = item;
        const edges = ssspAdjList[initial];

        if (edges) {
          for (const edge of edges) {
            this.step(edge, queue, ssspParent, ssspDistance);
          }
        }
      } else {
        expanded = item;
        this.step(expanded, queue, ssspParent, ssspDistance);
      }
    }

    console.log('Naïve Gabow:');
    console.log('Distances:', ssspDistance);
    console.log('Parents:', ssspParent);
    console.log();

    return ssspDistance;
  }

  bfsssp(source: number): number[] {
    const ssspParent = [...this.parent];
    const ssspDistance = [...this.distance];
    const ssspAdjList = this.adjList.map((edges) => (edges ? [...edges] : null));

    ssspDistance[source] = 0;

    for (let i = 0; i < ssspAdjList.length; i++) {
      const edges = ssspAdjList[i];

      if (edges) {
        for (let j = 0; j < edges.length; j++) {
          const edge = edges[j]!;
          const { weight, initial, terminal } = edge;

          if (ssspDistance[terminal] > ssspDistance[initial] + weight) {
            ssspDistance[terminal] = ssspDistance[initial] + weight;
            ssspParent[terminal] = initial;
          }
        }
      }
    }

    console.log('Bellman-Ford:');
    console.log('Distances:', ssspDistance);
    console.log('Parents:', ssspParent);
    console.log();

    return ssspDistance;
  }

  step(edge: Edge, queue: Array<number | Edge>, ssspParent: number[], ssspDistance: number[]): void {
    const { weight, initial, terminal, step } = edge;

    if (step > this.factor) {
      edge.step = step - 1;
      queue.push(edge);
    } else if (ssspDistance[terminal] > ssspDistance[initial] + weight) {
      if (ssspDistance[terminal] < this.absoluteMinimum) {
        ssspDistance[terminal] = this.NEG_INF;
      } else {
        ssspDistance[terminal] = ssspDistance[initial] + weight;
      }

      ssspParent[terminal] = initial;
      queue.push(terminal);
    }
  }

  toString(): string {
    return 'BFSE [adjList=' + JSON.stringify(this.adjList) + ']';
  }
}

// Example usage:
const bfs = new BFSE();
bfs.addEdgeWithParams(5, 0, 1);
bfs.addEdgeWithParams(-2, 1, 2);
bfs.addEdgeWithParams(10, 0, 2);
bfs.addEdgeWithParams(3, 2, 3);
bfs.sssp(0);
bfs.bfsssp(0);